lib/parsers/aac_parser/adts_header_info.rb

```diff
@@ -0,0 +1,134 @@
+# This is a representation of the relevant information found in an Audio Data Transport Stream (ADTS) file header.
+class FormatParser::AdtsHeaderInfo
+  attr_accessor :mpeg_version, :layer, :protection_absence, :profile, :mpeg4_sampling_frequency_index,
+                :mpeg4_channel_config, :originality, :home_usage, :frame_length, :buffer_fullness,
+                :aac_frames_per_adts_frame
+
+  # An ADTS header has the following format, when represented in bits:
+  # AAAAAAAA AAAABCCD EEFFFFGH HHIJKLMM MMMMMMMM MMMOOOOO OOOOOOPP (QQQQQQQQ QQQQQQQQ)
+  # The chunks represented by these letters have specific meanings, as described here:
+  # https://wiki.multimedia.cx/index.php/ADTS
+
+  AAC_ADTS_HEADER_BITS_CHUNK_SIZES = [['A', 12], ['B', 1], ['C', 2], ['D', 1], ['E', 2], ['F', 4], ['G', 1], ['H', 3], ['I', 1], ['J', 1], ['K', 1], ['L', 1], ['M', 13], ['O', 11], ['P', 2], ['Q', 16]]
+  MPEG4_AUDIO_OBJECT_TYPE_RANGE = 0..45
+  MPEG4_AUDIO_SAMPLING_FREQUENCY_RANGE = 0..14
+  MPEG4_AUDIO_SAMPLING_FREQUENCY_HASH = {
+    0 => 96000, 1 => 88200, 2 => 64000, 3 => 48000, 4 => 44100, 5 => 32000, 6 => 24000,
+    7 => 22050, 8 => 16000, 9 => 12000, 10 => 11025, 11 => 8000, 12 => 7350,
+    13 => 'Reserved', 14 => 'Reserved'
+  }
+  AAC_PROFILE_DESCRIPTION_HASH = { 0 => 'AAC_MAIN', 1 => 'AAC_LC (Low Complexity)', 2 => 'AAC_SSR (Scaleable Sampling Rate)', 3 => 'AAC_LTP (Long Term Prediction)' }
+  MPEG_VERSION_HASH = { 0 => 'MPEG-4', 1 => 'MPEG-2'}
+
+  def mpeg4_sampling_frequency
+    if !@mpeg4_sampling_frequency_index.nil? && MPEG4_AUDIO_SAMPLING_FREQUENCY_HASH.has_key?(@mpeg4_sampling_frequency_index)
+      return MPEG4_AUDIO_SAMPLING_FREQUENCY_HASH[@mpeg4_sampling_frequency_index]
+    end
+    nil
+  end
+
+  def profile_description
+    if !@profile.nil? && AAC_PROFILE_DESCRIPTION_HASH.has_key?(@profile)
+      return AAC_PROFILE_DESCRIPTION_HASH[@profile]
+    end
+    nil
+  end
+
+  def mpeg_version_description
+    if !@mpeg_version.nil? && MPEG_VERSION_HASH.has_key?(@mpeg_version)
+      return MPEG_VERSION_HASH[@mpeg_version]
+    end
+    nil
+  end
+
+  def number_of_audio_channels
+    case @mpeg4_channel_config
+    when 1..6
+      @mpeg4_channel_config
+    when 7
+      8
+    else
+      nil
+    end
+  end
+
+  def fixed_bitrate?
+    # A buffer fullness value of 0x7FF (decimal: 2047) denotes a variable bitrate, for which buffer fullness isn't applicable
+    @buffer_fullness != 2047
+  end
+
+  # The frame rate - i.e. frames per second
+  def frame_rate
+    # An AAC sample uncompresses to 1024 PCM samples
+    mpeg4_sampling_frequency.to_f / 1024
+  end
+
+  # If the given bit array is a valid ADTS header, this method will parse it and return an instance of AdtsHeaderInfo.
+  # Will return nil if the header does not match the ADTS specifications.
+  def self.parse_adts_header(header_bits)
```

Comment:
qq: why did you go for a static parser method instead of a constructor?

Reply:
I find this quite Ruby_esque_(_ful_?) 👍

Reply:
I wanted a clean parsing method that returned either the parsed object or a nil if the header was not ADTS compliant. If I used a constructor for this I'd probably have to throw a parsing error and deal with that. I'm open to suggestions though - what would make more sense from a Ruby dev point of view?

Reply:
It makes sense